V1.0.0 RC (#182)
OWASP/maswe: weaknesses/MASVS-RESILIENCE/MASWE-0061.md

Author: Sergio: DONE

---
title: Dynamic Analysis Tools Detection Not Implemented
id: MASWE-0061
alias: dynamic-analysis-tools

> requirement: "The app detects dynamic analysis tools."

Reviewer: ```suggestion
requirement: "The app detects dynamic analysis tools and responds to protect sensitive operations."
```

platform: [android, ios]
profiles: [R]
threat: MAS-THREAT-0061
attacks: [MAS-ATTACK-0003]
mappings:
  masvs-v1: [MSTG-RESILIENCE-4]
  masvs-v2: [MASVS-RESILIENCE-4]

>   cwe: [693]

Reviewer: ```suggestion
```

  maswe-beta: [MASWE-0102]

> status: new

Reviewer: ```suggestion
  maswe-beta: [MASWE-0102]
refs:
- https://frida.re/docs/home/
- https://github.com/LSPosed/LSPosed
- https://github.com/LSPosed/LSPlant
- https://www.cydiasubstrate.com/api/c/MSHookFunction/
- https://github.com/tealbathingsuit/ellekit
- https://github.com/roothide/Bootstrap
- https://burjcdigital.urjc.es/server/api/core/bitstreams/d249ebbb-5923-48ea-9215-af14cf2cc0b9/content
status: new
```

---

## Overview

This weakness occurs when an app does not detect the presence of dynamic instrumentation and hooking frameworks at runtime.

> Tools such as Frida, Xposed/LSPosed, and ElleKit/Cydia Substrate let attackers observe and modify the app while it runs, replacing method implementations and defeating client-side security controls. These frameworks leave detectable artifacts, such as loaded libraries, named pipes, listening ports, and installed hooks, that the app can check for. Without detection and a response, instrumentation proceeds unnoticed.

Reviewer: - Add note for common patching of instrumentation tools such as Florida Frida
- Add clarification to differentiate from MASWE-0064

```suggestion
Tools such as Frida, Xposed/LSPosed, and ElleKit/Cydia Substrate let attackers observe and modify the app while it runs, replacing method implementations and defeating client-side security controls. These frameworks can leave detectable artifacts, such as loaded libraries, named pipes, listening ports, installed hooks, or framework-specific classes, but attackers can often rename, relocate, hide, or patch these indicators to bypass shallow detection. Without layered detection and a response, instrumentation can proceed unnoticed.

Unlike runtime code integrity verification (see @MASWE-0064), this weakness focuses on detecting the presence of instrumentation tools and frameworks through tool-specific indicators. On the other hand, runtime code integrity verification checks whether loaded code, executable memory, or security-critical functions were modified, even when no known tool artifact is present.
```

## Modes of Introduction

- **No Instrumentation Checks**: Shipping without any detection of well-known instrumentation and hooking frameworks.

> - **Artifact Checks Missing or Shallow**: Not checking for hooking artifacts such as suspicious loaded libraries, named pipes, listening ports, or modified function prologues, or checking only from easily hooked Java/Swift code.

Reviewer: _"modified function prologues"_. We already have MASWE-0064 for code integrity, so we can orient that towards known function prologues such as frida's.

```suggestion
- **Artifact Checks Missing or Shallow**: Not checking for hooking artifacts such as suspicious loaded libraries, named pipes, listening ports, or known trampoline instruction sequences in function prologues, or checking only from easily hooked Java/Swift code.
```
